{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Check on the test and/or production databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv(override=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pypika import Query, Table, Field, Column, Criterion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.upsert import db_upsert\n",
    "from SRAgent.db.utils import db_list_tables, db_glimpse_tables, db_get_table, execute_query\n",
    "from SRAgent.db.get import db_find_srx\n",
    "from SRAgent.db.create import create_table, create_table_router"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "screcounter_star_results\n",
      "eval\n",
      "screcounter_trace\n",
      "srx_srr\n",
      "srx_metadata\n",
      "screcounter_log\n",
      "screcounter_star_params\n"
     ]
    }
   ],
   "source": [
    "# list all of the tables in prod\n",
    "os.environ['DYNACONF'] = 'prod'\n",
    "conn_prod = db_connect() \n",
    "print(\"\\n\".join(db_list_tables(conn_prod)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['screcounter_star_results',\n",
       " 'eval',\n",
       " 'screcounter_trace',\n",
       " 'srx_srr',\n",
       " 'srx_metadata',\n",
       " 'screcounter_log',\n",
       " 'screcounter_star_params']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_list_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- Table: screcounter_star_results --#\n",
      "sample\tfeature\testimated_number_of_cells\tfraction_of_unique_reads_in_cells\tmean_gene_per_cell\tmean_umi_per_cell\tmean_feature_per_cell\tmedian_gene_per_cell\tmedian_umi_per_cell\tmedian_feature_per_cell\tnumber_of_reads\tmedian_reads_per_cell\tq30_bases_in_cb_umi\tq30_bases_in_rna_read\treads_mapped_to_gene__unique_gene\treads_mapped_to_gene__unique_multiple_gene\treads_mapped_to_genefull__unique_genefull\treads_mapped_to_genefull__unique_multiple_genefull\treads_mapped_to_genefull_ex50pas__unique_genefull_ex50pas\treads_mapped_to_genefull_ex50pas__unique_multiple_genefull_ex50\treads_mapped_to_genefull_exonoverintron__unique_genefull_exonov\treads_mapped_to_genefull_exonoverintron__unique_multiple_genefu\treads_mapped_to_genome__unique\treads_mapped_to_genome__unique_multiple\treads_mapped_to_velocyto__unique_velocyto\treads_mapped_to_velocyto__unique_multiple_velocyto\treads_with_valid_barcodes\tsequencing_saturation\ttotal_feature_detected\tumis_in_cells\tunique_reads_in_cells_mapped_to_gene\tunique_reads_in_cells_mapped_to_genefull\tunique_reads_in_cells_mapped_to_genefull_ex50pas\tunique_reads_in_cells_mapped_to_genefull_exonoverintron\tcreated_at\tupdated_at\n",
      "ERX10024831\tgene_full\t2292\t0.911745\t\t12557.0\t2730.0\t\t10513.0\t2775.0\t146018831\t25399.0\t0.968155\t0.918635\t\t\t0.534665\t0.788618\t\t\t\t\t0.734504\t0.87767\t\t\t0.983305\t0.593299\t21731.0\t28780763\t\t71181072.0\t\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tgene_ex50\t2290\t0.912572\t\t17603.0\t2756.0\t\t11784.0\t2799.0\t146018831\t28689.0\t0.968155\t0.918635\t\t\t\t\t0.762455\t\t\t\t0.734504\t0.87767\t\t\t0.983334\t0.601129\t21911.0\t40311609\t\t\t101599277.0\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tgene_ex_int\t2290\t0.912561\t\t17625.0\t2767.0\t\t11803.0\t2809.0\t146018831\t28735.0\t0.968155\t0.918635\t\t\t\t\t\t\t\t\t0.734504\t0.87767\t\t\t0.983322\t0.601073\t22001.0\t40362108\t\t\t\t101712901.0\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10024831\tvelocyto\t0\t\t\t\t\t\t\t\t146018831\t\t0.968155\t0.918635\t\t\t\t\t\t\t\t\t0.734504\t0.87767\t0.0\t0.0\t0.984647\t-inf\t\t0\t\t\t\t\t2025-01-13 15:27:29.589416\t2025-01-14 20:06:25.557352\n",
      "ERX10086874\tgene_full\t2262\t0.909667\t\t4977.0\t2018.0\t\t4615.0\t2116.0\t160000000\t43880.0\t0.952429\t0.920767\t\t\t0.745753\t0.801599\t\t\t\t\t0.800461\t0.924752\t\t\t0.967224\t0.887163\t21453.0\t11258108\t\t108541887.0\t\t\t2025-01-13 15:17:54.767872\t2025-01-14 20:06:25.557352\n",
      "\n",
      "#-- Table: eval --#\n",
      "dataset_id\tdatabase\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\torganism\tcell_prep\tcreated_at\tupdated_at\n",
      "eval1\tsra\t29115023\tERX11157721\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tzebrafish\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t36178506\tERX11887200\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tmouse\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t5920965\tERX2188110\tyes\tyes\tyes\tother\tnot_applicable\tmacaque\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tsra\t34439895\tERX9692805\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tchicken\tsingle_cell\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "eval1\tgds\t200271457\tNaN\tno\tno\tno\tnot_applicable\tnot_applicable\thuman\tnot_applicable\t2025-01-03 18:38:01.789148\t2025-01-03 18:38:01.789148\n",
      "\n",
      "#-- Table: screcounter_trace --#\n",
      "task_id\thash\tnative_id\tname\tstatus\texit\tsubmit\tcontainer\tcpus\ttime\tdisk\tmemory\tattempt\tduration\trealtime\tcpu_percent\tpeak_rss\tpeak_vmem\trchar\twchar\tworkdir\tscratch\tcreated_at\tupdated_at\n",
      "1\t56/d8d487\tnf-56d8d487-1736775758365/0/nf-56d8d487-173677-6a5e0c92-ed1e-4c820\tDB_ACC_WF:GET_DB_ACCESSIONS\tCOMPLETED\t0\t2025-01-13 05:42:40.879\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 53s\t1.1s\t66.4%\t3 MB\t4.2 MB\t20.1 MB\t91.6 KB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/56/d8d4879ac276a9d30ec850f1ecf20a\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "2\te8/3237c0\tnf-e83237c0-1736775874725/0/nf-e83237c0-173677-e84d9f6b-20f7-48380\tSRA_STAT (1)\tCOMPLETED\t0\t2025-01-13 05:44:35.213\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t40.2%\t79.2 MB\t543.5 MB\t14.8 MB\t1.8 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/e8/3237c03758be58eca921e818fa3e0e\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "3\t64/3df771\tnf-643df771-1736775874726/0/nf-643df771-173677-54f5b596-489b-43870\tSRA_STAT (2)\tCOMPLETED\t0\t2025-01-13 05:44:35.674\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t37.7%\t79.5 MB\t543.5 MB\t14.8 MB\t755 KB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/64/3df7716232b892d0d8a7fa72ece0b8\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "4\t4e/793c08\tnf-4e793c08-1736775874789/0/nf-4e793c08-173677-239dfe74-6cc5-41130\tSRA_STAT (3)\tCOMPLETED\t0\t2025-01-13 05:44:36.077\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 39s\t1.3s\t39.1%\t79.4 MB\t543.5 MB\t14.8 MB\t1.7 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/4e/793c0898e1f5b974d97dca646c8fa2\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "5\tce/e8e21e\tnf-cee8e21e-1736775874795/0/nf-cee8e21e-173677-f6013cb6-ca08-40df0\tSRA_STAT (4)\tCOMPLETED\t0\t2025-01-13 05:44:36.533\tus-east1-docker.pkg.dev/c-tc-429521/sc-recounter-download/sc-recounter-download:0.1.0\t1\t2h\t50 GB\t4 GB\t1\t1m 38s\t1.3s\t39.7%\t79.1 MB\t543.5 MB\t14.8 MB\t1.7 MB\tgs://arc-ctc-nextflow/scRecounter/prod/work/SCRECOUNTER_2025-01-13_05-42-15/ce/e8e21e2d91358d48f9b67118002b20\ttrue\t2025-01-13 16:13:10.448466\t2025-01-13 16:13:10.448466\n",
      "\n",
      "#-- Table: srx_srr --#\n",
      "srx_accession\tsrr_accession\tcreated_at\tupdated_at\n",
      "SRX24914484\tSRR29400843\t2025-01-03 18:43:49.275138\t2025-01-03 18:43:49.275138\n",
      "SRX22006379\tSRR26297925\t2025-01-03 18:43:57.864762\t2025-01-03 18:43:57.864762\n",
      "SRX20273630\tSRR24488012\t2025-01-03 18:44:17.303662\t2025-01-03 18:44:17.303662\n",
      "SRX25634198\tSRR30167498\t2025-01-03 18:44:30.759838\t2025-01-03 18:44:30.759838\n",
      "SRX25602002\tSRR30132651\t2025-01-03 18:45:27.695421\t2025-01-03 18:45:27.695421\n",
      "\n",
      "#-- Table: srx_metadata --#\n",
      "database\tentrez_id\tsrx_accession\tis_illumina\tis_single_cell\tis_paired_end\tlib_prep\ttech_10x\tcell_prep\torganism\ttissue\tdisease\tpurturbation\tcell_line\tczi_collection_id\tczi_collection_name\tnotes\tcreated_at\tupdated_at\n",
      "sra\t26032408\tERX8773994\tyes\tyes\tyes\tSmart-seq2\tnot_applicable\tsingle_cell\tmouse\tspinal cord\tnone\tTg(FOXJ1-EGFP)85Leo transgenic mice\tependymal cell\t\t\tMetadata obtained by SRAgent\t2025-01-21 01:48:26.414116\t2025-01-21 01:51:22.064081\n",
      "sra\t26580801\tSRX19329973\tyes\tyes\tyes\t10x_Genomics\tatac\tsingle_nucleus\thuman\tleft ventricle\tAtrial Fibrillation (AF)\tunsure\tunsure\t20eea6c8-9d64-42c9-9b6f-c11b5249e0e9\tSingle-cell genomics improves the discovery of risk variants and genes of atrial fibrillation\tMetadata obtained by SRAgent\t2025-01-03 18:39:40.221883\t2025-02-14 07:27:49.489018\n",
      "sra\t14870188\tSRX11169657\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tsingle_cell\thuman\tlung\tnormal\tnot specified\tairway cells\tf86d6317-7215-409e-bfda-3f4ded3dadaa\tHuman CellCards Multi-Study CellRef 1.0 Atlas\tMetadata obtained by SRAgent\t2025-01-03 18:39:40.221883\t2025-02-14 23:16:06.463386\n",
      "sra\t9287471\tSRX7063667\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tsingle_cell\thuman\tbrain tumor\tglioblastoma multiforme\tunsure\tfreshly sorted glioblastoma cells; CD133/GLAST\t999f2a15-3d7e-440b-96ae-2c806799c08c\tHarmonized single-cell landscape, intercellular crosstalk and tumor architecture of glioblastoma\tMetadata obtained by SRAgent\t2025-01-03 18:39:40.221883\t2025-02-15 00:01:04.661922\n",
      "sra\t18806354\tSRX13549220\tyes\tyes\tyes\t10x_Genomics\t3_prime_gex\tsingle_cell\tGallus gallus\tliver\tunsure\tunsure\tunsure\t74e10dc4-cbb2-4605-a189-8a1cd8e44d8c\tSpatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches\tMetadata obtained by SRAgent\t2025-01-03 18:39:40.221883\t2025-02-15 01:32:48.044914\n",
      "\n",
      "#-- Table: screcounter_log --#\n",
      "sample\taccession\tprocess\tstep\tstatus\tmessage\tcreated_at\tupdated_at\n",
      "ERX10024831\tERR10509460\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10024831\tERR10509461\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10086874\tERR10614597\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10086874\tERR10614598\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "ERX10335126\tERR10891936\tGet db accessions\tFinal\tSuccess\tObtained database accession for processing\t2025-01-13 13:44:01.221104\t2025-01-13 13:44:01.221104\n",
      "\n",
      "#-- Table: screcounter_star_params --#\n",
      "sample\tbarcodes\tstar_index\tcell_barcode_length\tumi_length\tstrand\tcreated_at\tupdated_at\n",
      "ERX10086874\t737K-august-2016.txt\tstar_refData_2020_mm10\t16\t10\tForward\t2025-01-13 13:58:53.927797\t2025-01-13 13:58:53.927797\n",
      "ERX10396585\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:56.112918\t2025-01-13 13:58:56.112918\n",
      "ERX10396584\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:58.510780\t2025-01-13 13:58:58.510780\n",
      "ERX10690512\t3M-february-2018.txt\tstar_refData_2020_mm10\t16\t12\tForward\t2025-01-13 13:58:58.640112\t2025-01-13 13:58:58.640112\n",
      "ERX10855739\t737K-august-2016.txt\tstar_refData_2020_hg38\t16\t10\tForward\t2025-01-13 13:58:59.433825\t2025-01-13 13:58:59.433825\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_glimpse_tables(conn_prod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Database updates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CZI datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Organism "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>is_illumina</th>\n",
       "      <th>is_single_cell</th>\n",
       "      <th>is_paired_end</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>cell_prep</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "      <th>purturbation</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "      <th>created_at</th>\n",
       "      <th>updated_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>5980061</td>\n",
       "      <td>SRX4401543</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>human</td>\n",
       "      <td>Human prostate basal epithelia</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>D4PrF_BE, CD326+/CD271+/CD26-</td>\n",
       "      <td>e2a4a67f-6a18-431a-ab9c-6e77dd31cc80</td>\n",
       "      <td>A Cellular Anatomy of the Normal Adult Human P...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-18 01:17:18.338743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>19008386</td>\n",
       "      <td>SRX13671170</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_nucleus</td>\n",
       "      <td>human</td>\n",
       "      <td>Dorsolateral Prefrontal cortex (Brodmann area 46)</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-18 01:48:15.592726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>21270522</td>\n",
       "      <td>ERX8792046</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Lung-draining lymph node</td>\n",
       "      <td>Prostate cancer</td>\n",
       "      <td>Saquinavir-NO treatment</td>\n",
       "      <td>PC-3</td>\n",
       "      <td>62ef75e4-cbea-454e-a0ce-998ec40223d3</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-16 03:08:04.360701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>33748102</td>\n",
       "      <td>ERX12098959</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>intercostal muscle</td>\n",
       "      <td>intracranial haemorrhage</td>\n",
       "      <td>unsure</td>\n",
       "      <td>nuclei from human skeletal muscle biopsies</td>\n",
       "      <td>2d40e6a7-f2fd-49ba-9db9-6b97e4c6dad5</td>\n",
       "      <td>Human skeletal muscle ageing atlas</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-15 15:20:42.842344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>30534579</td>\n",
       "      <td>SRX22540304</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Peripheral Blood Mononuclear Cells (PBMCs)</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>COVID-19 mRNA vaccine (BNT162b2)</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>ecb739c5-fe0d-4b48-81c6-217c4d64eec4</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-15 23:28:22.014208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6055</th>\n",
       "      <td>sra</td>\n",
       "      <td>30533995</td>\n",
       "      <td>SRX22539720</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>5_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>COVID-19 vaccination</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>ecb739c5-fe0d-4b48-81c6-217c4d64eec4</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-17 01:40:39.811736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6056</th>\n",
       "      <td>sra</td>\n",
       "      <td>19008361</td>\n",
       "      <td>SRX13671145</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_nucleus</td>\n",
       "      <td>human</td>\n",
       "      <td>Dorsolateral Prefrontal Cortex (Brodmann area 46)</td>\n",
       "      <td>Alzheimer's Disease</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>91c8e321-566f-4f9d-b89e-3a164be654d5</td>\n",
       "      <td>Neuron type-specific effects of human aging an...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-17 03:59:44.582294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6057</th>\n",
       "      <td>sra</td>\n",
       "      <td>5987222</td>\n",
       "      <td>SRX4405260</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>MARS-seq</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Multiple myeloma (MM)</td>\n",
       "      <td>Untreated</td>\n",
       "      <td>Control donor (AB2835)</td>\n",
       "      <td>2a0b02c0-fea6-47bd-92b9-9b03f5d2580c</td>\n",
       "      <td>Single cell dissection of plasma cell heteroge...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-15 14:20:46.277606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6058</th>\n",
       "      <td>sra</td>\n",
       "      <td>9278658</td>\n",
       "      <td>SRX7059587</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>brain</td>\n",
       "      <td>glioblastoma, stage IV</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>999f2a15-3d7e-440b-96ae-2c806799c08c</td>\n",
       "      <td>Harmonized single-cell landscape, intercellula...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-15 03:37:30.793431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6059</th>\n",
       "      <td>sra</td>\n",
       "      <td>11941283</td>\n",
       "      <td>SRX9165759</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>other</td>\n",
       "      <td>single_cell</td>\n",
       "      <td>human</td>\n",
       "      <td>Fetal Small Intestine</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>unsure</td>\n",
       "      <td>60358420-6055-411d-ba4f-e8ac80682a2e</td>\n",
       "      <td>Spatiotemporal analysis of human intestinal de...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "      <td>2025-01-03 18:39:40.221883</td>\n",
       "      <td>2025-02-16 05:35:58.741302</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6060 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     database  entrez_id srx_accession is_illumina is_single_cell  \\\n",
       "0         sra    5980061    SRX4401543         yes             no   \n",
       "1         sra   19008386   SRX13671170         yes            yes   \n",
       "2         sra   21270522    ERX8792046         yes            yes   \n",
       "3         sra   33748102   ERX12098959         yes            yes   \n",
       "4         sra   30534579   SRX22540304         yes            yes   \n",
       "...       ...        ...           ...         ...            ...   \n",
       "6055      sra   30533995   SRX22539720         yes            yes   \n",
       "6056      sra   19008361   SRX13671145         yes             no   \n",
       "6057      sra    5987222    SRX4405260         yes            yes   \n",
       "6058      sra    9278658    SRX7059587         yes            yes   \n",
       "6059      sra   11941283    SRX9165759         yes            yes   \n",
       "\n",
       "     is_paired_end      lib_prep        tech_10x       cell_prep organism  \\\n",
       "0               no         other  not_applicable  not_applicable    human   \n",
       "1              yes         other  not_applicable  single_nucleus    human   \n",
       "2               no  10x_Genomics     5_prime_gex     single_cell    human   \n",
       "3              yes  10x_Genomics     3_prime_gex     single_cell    human   \n",
       "4              yes  10x_Genomics     5_prime_gex     single_cell    human   \n",
       "...            ...           ...             ...             ...      ...   \n",
       "6055           yes  10x_Genomics     5_prime_gex     single_cell    human   \n",
       "6056           yes         other  not_applicable  single_nucleus    human   \n",
       "6057            no      MARS-seq  not_applicable     single_cell    human   \n",
       "6058           yes  10x_Genomics     3_prime_gex     single_cell    human   \n",
       "6059           yes  10x_Genomics           other     single_cell    human   \n",
       "\n",
       "                                                 tissue  \\\n",
       "0                        Human prostate basal epithelia   \n",
       "1     Dorsolateral Prefrontal cortex (Brodmann area 46)   \n",
       "2                              Lung-draining lymph node   \n",
       "3                                    intercostal muscle   \n",
       "4            Peripheral Blood Mononuclear Cells (PBMCs)   \n",
       "...                                                 ...   \n",
       "6055                                              PBMCs   \n",
       "6056  Dorsolateral Prefrontal Cortex (Brodmann area 46)   \n",
       "6057                                        Bone Marrow   \n",
       "6058                                              brain   \n",
       "6059                              Fetal Small Intestine   \n",
       "\n",
       "                       disease                      purturbation  \\\n",
       "0                       unsure                            unsure   \n",
       "1                       unsure                            unsure   \n",
       "2              Prostate cancer           Saquinavir-NO treatment   \n",
       "3     intracranial haemorrhage                            unsure   \n",
       "4                     COVID-19  COVID-19 mRNA vaccine (BNT162b2)   \n",
       "...                        ...                               ...   \n",
       "6055                  COVID-19              COVID-19 vaccination   \n",
       "6056       Alzheimer's Disease                            unsure   \n",
       "6057     Multiple myeloma (MM)                         Untreated   \n",
       "6058    glioblastoma, stage IV                            unsure   \n",
       "6059                    unsure                            unsure   \n",
       "\n",
       "                                       cell_line  \\\n",
       "0                  D4PrF_BE, CD326+/CD271+/CD26-   \n",
       "1                                         unsure   \n",
       "2                                           PC-3   \n",
       "3     nuclei from human skeletal muscle biopsies   \n",
       "4                                          PBMCs   \n",
       "...                                          ...   \n",
       "6055                                       PBMCs   \n",
       "6056                                      unsure   \n",
       "6057                      Control donor (AB2835)   \n",
       "6058                                      unsure   \n",
       "6059                                      unsure   \n",
       "\n",
       "                         czi_collection_id  \\\n",
       "0     e2a4a67f-6a18-431a-ab9c-6e77dd31cc80   \n",
       "1     91c8e321-566f-4f9d-b89e-3a164be654d5   \n",
       "2     62ef75e4-cbea-454e-a0ce-998ec40223d3   \n",
       "3     2d40e6a7-f2fd-49ba-9db9-6b97e4c6dad5   \n",
       "4     ecb739c5-fe0d-4b48-81c6-217c4d64eec4   \n",
       "...                                    ...   \n",
       "6055  ecb739c5-fe0d-4b48-81c6-217c4d64eec4   \n",
       "6056  91c8e321-566f-4f9d-b89e-3a164be654d5   \n",
       "6057  2a0b02c0-fea6-47bd-92b9-9b03f5d2580c   \n",
       "6058  999f2a15-3d7e-440b-96ae-2c806799c08c   \n",
       "6059  60358420-6055-411d-ba4f-e8ac80682a2e   \n",
       "\n",
       "                                    czi_collection_name  \\\n",
       "0     A Cellular Anatomy of the Normal Adult Human P...   \n",
       "1     Neuron type-specific effects of human aging an...   \n",
       "2     Cross-tissue immune cell analysis reveals tiss...   \n",
       "3                    Human skeletal muscle ageing atlas   \n",
       "4     COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "...                                                 ...   \n",
       "6055  COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "6056  Neuron type-specific effects of human aging an...   \n",
       "6057  Single cell dissection of plasma cell heteroge...   \n",
       "6058  Harmonized single-cell landscape, intercellula...   \n",
       "6059  Spatiotemporal analysis of human intestinal de...   \n",
       "\n",
       "                             notes                 created_at  \\\n",
       "0     Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "1     Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "2     Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "3     Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "4     Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "...                            ...                        ...   \n",
       "6055  Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "6056  Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "6057  Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "6058  Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "6059  Metadata obtained by SRAgent 2025-01-03 18:39:40.221883   \n",
       "\n",
       "                     updated_at  \n",
       "0    2025-02-18 01:17:18.338743  \n",
       "1    2025-02-18 01:48:15.592726  \n",
       "2    2025-02-16 03:08:04.360701  \n",
       "3    2025-02-15 15:20:42.842344  \n",
       "4    2025-02-15 23:28:22.014208  \n",
       "...                         ...  \n",
       "6055 2025-02-17 01:40:39.811736  \n",
       "6056 2025-02-17 03:59:44.582294  \n",
       "6057 2025-02-15 14:20:46.277606  \n",
       "6058 2025-02-15 03:37:30.793431  \n",
       "6059 2025-02-16 05:35:58.741302  \n",
       "\n",
       "[6060 rows x 19 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in srx-metadata as pandas dataframe\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .select(tbl.star) \\\n",
    "    .distinct() \\\n",
    "    .where(tbl.czi_collection_id.isnull().negate())\n",
    "srx_metadata_cxg = pd.read_sql(str(stmt), conn_prod)\n",
    "srx_metadata_cxg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "organism\n",
       "human                4990\n",
       "mouse                1005\n",
       "Macaca mulatta         36\n",
       "NaN                    14\n",
       "Rattus norvegicus       5\n",
       "Sus scrofa              4\n",
       "Danio rerio             3\n",
       "Gallus gallus           2\n",
       "other                   1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# summarize organism\n",
    "srx_metadata_cxg.groupby(\"organism\").size().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3641</th>\n",
       "      <td>SRX13549222</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     srx_accession organism\n",
       "3641   SRX13549222    other"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter to just `other` organisms\n",
    "srx_metadata_cxg[srx_metadata_cxg[\"organism\"] == \"other\"][[\"srx_accession\", \"organism\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX9556667</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX12060278</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERX12060274</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERX12060276</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERX12060273</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERX12060297</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERX12060299</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERX12060275</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRX13549222</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRX9556651</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERX12060277</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   srx_accession organism\n",
       "0     SRX9556667    human\n",
       "1    ERX12060278    human\n",
       "2    ERX12060274    human\n",
       "3    ERX12060276    human\n",
       "4    ERX12060273    human\n",
       "5    ERX12060297    human\n",
       "6    ERX12060299    human\n",
       "7    ERX12060275    human\n",
       "8    SRX13549222    other\n",
       "9     SRX9556651    human\n",
       "10   ERX12060277    human"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # update \"other\" organisms to correct\n",
    "# idx = [\n",
    "#     ['SRX9556667', 'human'], \n",
    "#     ['ERX12060278', 'human'], \n",
    "#     ['ERX12060274', 'human'], \n",
    "#     ['ERX12060276', 'human'], \n",
    "#     ['ERX12060273', 'human'], \n",
    "#     ['ERX12060297', 'human'], \n",
    "#     ['ERX12060299', 'human'], \n",
    "#     ['ERX12060275', 'human'], \n",
    "#     ['SRX13549222', 'other'],\n",
    "#     ['SRX9556651', 'human'], \n",
    "#     ['ERX12060277', 'human'], \n",
    "# ]\n",
    "# idx = pd.DataFrame(idx, columns=[\"srx_accession\", \"organism\"])\n",
    "# idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488082</td>\n",
       "      <td>SRX9556667</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sra</td>\n",
       "      <td>32114110</td>\n",
       "      <td>ERX12060278</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sra</td>\n",
       "      <td>32114006</td>\n",
       "      <td>ERX12060274</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sra</td>\n",
       "      <td>32113144</td>\n",
       "      <td>ERX12060276</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sra</td>\n",
       "      <td>32114450</td>\n",
       "      <td>ERX12060273</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sra</td>\n",
       "      <td>32114882</td>\n",
       "      <td>ERX12060297</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sra</td>\n",
       "      <td>32113802</td>\n",
       "      <td>ERX12060299</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sra</td>\n",
       "      <td>32113700</td>\n",
       "      <td>ERX12060275</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sra</td>\n",
       "      <td>18806356</td>\n",
       "      <td>SRX13549222</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sra</td>\n",
       "      <td>12488066</td>\n",
       "      <td>SRX9556651</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sra</td>\n",
       "      <td>32113404</td>\n",
       "      <td>ERX12060277</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   database  entrez_id srx_accession organism\n",
       "0       sra   12488082    SRX9556667    human\n",
       "1       sra   32114110   ERX12060278    human\n",
       "2       sra   32114006   ERX12060274    human\n",
       "3       sra   32113144   ERX12060276    human\n",
       "4       sra   32114450   ERX12060273    human\n",
       "5       sra   32114882   ERX12060297    human\n",
       "6       sra   32113802   ERX12060299    human\n",
       "7       sra   32113700   ERX12060275    human\n",
       "8       sra   18806356   SRX13549222    other\n",
       "9       sra   12488066    SRX9556651    human\n",
       "10      sra   32113404   ERX12060277    human"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# with with other columns\n",
    "srx_metadata_cxg_f = pd.merge(srx_metadata_cxg.drop(columns=[\"organism\"]), idx, on=\"srx_accession\", how=\"inner\")\n",
    "srx_metadata_cxg_f = srx_metadata_cxg_f[[\"database\", \"entrez_id\", \"srx_accession\", \"organism\"]]\n",
    "srx_metadata_cxg_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update records\n",
    "from SRAgent.db.connect import db_connect\n",
    "from SRAgent.db.update import db_update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "with db_connect() as conn:\n",
    "    db_update(srx_metadata_cxg_f, \"srx_metadata\", conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# --OLD--"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out records with no SRX accessions\n",
    "\n",
    "`New dataset found by Find-Datasets agent`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entrez_ids = srx_metadata[srx_metadata[\"srx_accession\"].isna()][\"entrez_id\"].unique().tolist()\n",
    "entrez_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete extrez_ids from srx_metadata\n",
    "tbl = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .delete() \\\n",
    "    .where(tbl.entrez_id.isin(entrez_ids))\n",
    "with conn_prod.cursor() as cur:\n",
    "    cur.execute(str(stmt))\n",
    "    conn_prod.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete from log\n",
    "tbl = Table(\"screcounter_log\")\n",
    "to_rm = [\"ERX11146221\"]\n",
    "stmt = Query \\\n",
    "    .from_(tbl) \\\n",
    "    .delete() \\\n",
    "    .where(tbl.sample.isin(to_rm))\n",
    "with conn_prod.cursor() as cur:\n",
    "    cur.execute(str(stmt))\n",
    "    conn_prod.commit()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
